crates/storage/derive/src/lib.rs

```diff
@@ -132,6 +134,44 @@ synstructure::decl_derive!(
     /// ```
     packed_layout_derive
 );
+
+synstructure::decl_derive!(
+    [SpreadAllocate] =>
+    /// Derives `ink_storage`'s `SpreadAllocate` trait for the given `struct`.
+    ///
+    /// # Note
+    ///
+    /// As of now `enum` types are not supported!
```

Comment:
SpreadAllocate is very similar to the Default trait. Both do default initialization of some kind. There is no trivial way to figure out the default variant of an enum.

Reply:
Would it be possible to implement `SpreadAllocate` for enums that implement the `Default` trait?

e.g. something like:

```
#[derive(Debug)]
pub enum Color {
    Red,
    Blue,
    Green,
    NoColor
}

impl Default for Color {
    fn default() -> Self { Color::NoColor }
}
```

I have a contract that has been making use of enums but I'm no longer able to compile it due to the new `initialize_contract` concept:

```
354 |             ink_lang::codegen::initialize_contract(|contract| {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `SpreadAllocate` is not implemented for `Prosopo`
    |
note: required by a bound in `initialize_contract`
   --> /home/user/.cargo/git/checkouts/ink-e2bdf0affead2725/4391ae5/crates/lang/src/codegen/dispatch/execution.rs:142:33
    |
142 |     Contract: ContractRootKey + SpreadAllocate,
    |                                 ^^^^^^^^^^^^^^ required by this bound in `initialize_contract`

```

Reply:
It would definitely make sense to allow `derive` impls for `enum`s with `Default` impl. Not sure how to implement this properly though. Maybe we could add a `Self: Default` `where` bound to the generated derive impl if the derive item is an `enum`:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=21d9e01c332a660d9e5ab22634250f1a

Reply:
I [modified](https://github.com/prosopo-io/ink/commit/bdc676d61804a3fd88f76cb8045238250b3db166) ink to allow me to pass an `enum` with `Default` as per your suggestion. I'm now able to use the `SpreadAllocate` derive function on my structs that contain enums.

```
    #[derive(Debug, SpreadLayout, SpreadAllocate)]
    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo, StorageLayout))]
    pub struct ProviderAccountMap(Mapping<AccountId, Provider>); //Provider contains a multi value enum

    impl ProviderAccountMap
    {
        fn insert(&mut self, key: AccountId, value: Provider) {
            self.0.insert(key, &value);
        }
        fn get(&self, key: &AccountId) -> Option<Provider> {
            return self.0.get(key);
        }
    }
```


Unfortunately, I wasn't able to generalise this for `T` as `scale_info::TypeInfo` is not available within ink in `no_std` mode. I was trying to do something like this:

```
    pub struct AccountMap<T>(Mapping<AccountId, T>)
        where T: scale::Encode + scale::Decode + SpreadLayout + PackedLayout + scale_info::TypeInfo;
    
    impl<T> AccountMap<T>
        where
            T: PackedLayout + scale::Encode + scale::Decode + scale::EncodeLike + scale_info::TypeInfo
    {
        fn insert(&mut self, key: AccountId, value: T) {
            self.0.insert(key, &value);
        }
        fn get(&self, key: &AccountId) -> Option<T> {
            return self.0.get(key);
        }
    }
```

Is this possible within the smart contract or am I wasting my time?

Reply:
> 

Could you please open an issue in order to resolve this?
I find it very unpleasant to talk about this feature request in a merged PR discussion.

Reply:
Done: https://github.com/paritytech/ink/issues/1042